src/PromiseInterface.php

```diff
@@ -32,10 +33,11 @@ interface PromiseInterface
      *      than once.
      *  3. `$onProgress` (deprecated) may be called multiple times.
      *
-     * @param callable|null $onFulfilled
+     * @template TReturn of mixed
+     * @param callable(T): TReturn $onFulfilled
      * @param callable|null $onRejected
      * @param callable|null $onProgress This argument is deprecated and should not be used anymore.
-     * @return PromiseInterface
+     * @return (TReturn is PromiseInterface ? TReturn : PromiseInterface<TReturn>)
      */
     public function then(callable $onFulfilled = null, callable $onRejected = null, callable $onProgress = null);
```

Comment:
I see this currently only affects the `then()` method, should we also add this for the methods defined in `ExtendedPromiseInterface`? (Despite being removed in Promise v3 via #75.)

Reply:
This is purposely only covering `then` and not the other methods. In #223 I want to expand that to other methods. But the primary goal is to get this ready for https://github.com/reactphp/async/pull/40 and the minimum we need for that is support only on the `then` method. If we can add more methods before that release of that I'm all up for it. But if we don't I propose we leave it out.